flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/resources/ClusterResourceManager.java

```diff
@@ -0,0 +1,282 @@
+ * cluster is using the Kubernetes Cluster Autoscaler, we will use this data to extrapolate the
+ * number of nodes to the maximum defined nodes in the autoscaler configuration.
+ *
+ * <p>We currently track CPU and memory. Ephemeral storage is missing because there is no easy way
+ * to get node statics on free storage.
+ */
+public class ClusterResourceManager implements ResourceCheck {
+
+    private static final Logger LOG = LoggerFactory.getLogger(ClusterResourceManager.class);
+
+    /** ConfigMap name of the Kubernetes Cluster Autoscaler. */
+    static final String CLUSTER_AUTOSCALER_CONFIG_MAP = "cluster-autoscaler-status";
+
+    /** EKS specific node group information. Code still works without this label. */
+    static final String LABEL_NODE_GROUP = "eks.amazonaws.com/nodegroup";
+
+    private final Duration refreshInterval;
+    private final KubernetesClient kubernetesClient;
+
+    @VisibleForTesting ClusterResourceView clusterResourceView;
+
+    public ClusterResourceManager(Duration refreshInterval, KubernetesClient kubernetesClient) {
+        this.refreshInterval = refreshInterval;
+        this.kubernetesClient = kubernetesClient;
+    }
+
+    @Override
+    public synchronized boolean trySchedule(
+            int currentInstances,
+            int newInstances,
+            double cpuPerInstance,
+            double memoryPerInstance) {
+
+        if (refreshInterval.isNegative()) {
+            // Feature disabled
+            return true;
+        }
+
+        if (shouldRefreshView(clusterResourceView, refreshInterval)) {
+            try {
+                clusterResourceView = createResourceView(kubernetesClient);
+            } catch (KubernetesClientException e) {
+                if (e.getCode() == 403) {
+                    LOG.warn(
+                            "No permission to retrieve node resource usage. Resource check disabled.");
+                    return true;
+                }
+                throw e;
+            }
+        }
+
+        if (newInstances <= currentInstances) {
+            LOG.debug("Fewer or same amount of resources used after scaling.");
+            return true;
+        }
+
+        return trySchedule(
+                clusterResourceView,
+                currentInstances,
+                newInstances,
+                cpuPerInstance,
+                memoryPerInstance);
+    }
+
+    /**
+     * Simple check whether the new resource requirements can be scheduled. Note: This is not a
+     * full-blown scheduler. We may return false-negatives, i.e. we may indicate scheduling is not
+     * possible when it actually is. This is still better than false positives where we suffer from
+     * downtime due to non-schedulable pods.
+     */
+    private static boolean trySchedule(
+            ClusterResourceView resourceView,
+            int currentInstances,
+            int newInstances,
+            double cpuPerInstance,
+            double memoryPerInstance) {
```

Comment:
Instead of having `current` and `newInstances` can we simply pass the diff / increment? That would make for a simpler logic and be more intuitive. Such as checking whether we can schedule 2 more instances

Reply:
I prefer not to because the scheduling is supposed to remove all instances from the internal resource view and then re-acquire them. Similarly to how the scaling works. Otherwise we end up with fragmentation issues and we would also be unable to recover from scenarios where we are overbooked. Providing more information will also allow us to  dynamically change the scheduling behavior.